shell/platform/embedder/embedder.cc

```diff
@@ -1746,3 +1747,17 @@ FlutterEngineResult FlutterEngineNotifyLowMemoryWarning(
                    kInternalInconsistency,
                    "Could not dispatch the low memory notification message.");
 }
+
+FlutterEngineResult FlutterEngineNotifyTimezoneUpdate(
+    FLUTTER_API_SYMBOL(FlutterEngine) engine) {
+  // Currently there is no shell specific timezone data since all we call is
+  // tzset. But we may want to store this information in the future and it would
+  // be too late then to change the API.
+#if OS_WIN
+  _tzset();
+#else
+  tzset();
```

Comment:
This is an implementation detail of how Dart determines the local time zone for `DateTime` objects.  I don't think Dart guarantees that it will always use this mechanism or that it won't cache the time zone in ways that make this API ineffective.

I'd recommend that applications use a library like https://pub.dev/packages/timezone if they need tighter control over time zones.  If we are going to add an API like this to the Flutter embedder, then I'd prefer that Dart add an API that Flutter can wrap.

Reply:
> This is an implementation detail of how Dart determines the local time zone for DateTime objects.

Thats a fair point but embedders today need to work around this implementation detail by manually calling `tzset`. This dependence on an unstable implementation detail currently resides in the embedder. This patch brings it closer to the engine. If the Dart VM makes changes to the API in the future, we can accommodate this fix for all embedders instead of chasing all existing implementations.

Reply:
> I'd recommend that applications use a library like https://pub.dev/packages/timezone if they need tighter control over time zones. If we are going to add an API like this to the Flutter embedder, then I'd prefer that Dart add an API that Flutter can wrap.

cc @rmacnak-google @a-siva who can probably speak more on how to go about achieving this.

Reply:
tzset is provided by the underlying system, it is not something the Dart VM provides.

Do we have a link to the original issue that triggered this change.

Reply:
b/145541732 is the internal issue. This was necessary because DateTime.now() would return incorrect results after a timezone update. We had to reverse engineer what the VM was doing and explicitly call `tzset()` in the embedder in order for new calls DateTime.now() to return correct results.

Reply:
As discussed with @a-siva offline, I was filing an issue for this. But, I [found a comment on a closed issue](https://github.com/dart-lang/sdk/issues/39436#issuecomment-555423012) that basically says this is working as intended and `DateTime.now()` is an unreliable way to get time in the local timezone.